parasol/schema.py

```diff
@@ -0,0 +1,251 @@
+    a :class:`SolrSchema`instance.
+    """
+    def __init__(self, field_class, analyzer):
+        self.field_class = field_class
+        self.analyzer = analyzer
+
+    def __get__(self, obj, objtype):
+        # return format neded for declaring field type
+        return {
+            'class': self.field_class,
+            'analyzer': self.analyzer.as_solr_config()
+        }
+
+    def __set__(self, obj, val):
+        # enforce read-only descriptor
+        raise AttributeError
+
+
+class SolrSchema:
+    '''Solr schema configuration'''
+
+    #: dictionary of copy fields to be configured
+    #: key is source field, value is destination field or list of fields
+    copy_fields = {}
+
+    @classmethod
+    def get_configuration(cls):
+        '''Find a SolrSchema subclass for use as schema configuration.
+        Currently only supports one schema configuration.'''
+        subclasses = cls.__subclasses__()
+        if not subclasses:
+            raise Exception('No Solr schema configuration found')
+        elif len(subclasses) > 1:
+            raise Exception('Currently only one Solr schema configuration is supported (found %d)' \
+                             % len(subclasses))
+
+        return subclasses[0]
+
+    @classmethod
+    def get_field_names(cls):
+        '''iterate over class attributes and return all that are instances of
+        :class:`SolrField`'''
+        return [attr_name for attr_name, attr_type in cls.__dict__.items()
+                if isinstance(attr_type, SolrField)]
+
+    @classmethod
+    def get_field_types(cls):
+        '''iterate over class attributes and return all that are instances of
+        :class:`SolrFieldType`'''
+        return [attr_name for attr_name, attr_type in cls.__dict__.items()
+                if isinstance(attr_type, SolrFieldType)]
+
+    @classmethod
+    def configure_fields(cls, solr):
+        '''Update the configured Solr instance schema to match
+        the configured fields.
+
+        Calls :meth:`configure_copy_fields` after
+        new fields have been created and before old fields are removed,
+        since an outdated copy field could prevent removal.
+
+        returns: :class:`attrdict.AttrDefault` with counts for added,
+        updated, and deleted fields.
+        '''
+
+        current_fields = [field.name for field in solr.schema.list_fields()]
+        configured_field_names = cls.get_field_names()
+
+        # use attrdict instead of defaultdict since we have attrmap
+        stats = AttrDefault(int, {})
+
+        for field_name in configured_field_names:
+            field_opts = getattr(cls, field_name)
+            if field_name not in current_fields:
+                logger.debug('Adding schema field %s %s', field_name, field_opts)
+                solr.schema.add_field(name=field_name, **field_opts)
+                stats.added += 1
+            else:
+                # NOTE: currently no check if field configuration has changed
```

Comment:
Checking that the field exists isn't hard - it's checking if the field exists with the _correct configuration_ that's difficult (not because I can't get it, just because comparing dicts is apparently hard). I thought we could do without it at this point and just always update in case of changes.

Is the comment unclear? Can you suggest how tor revise?

Reply:
`NOTE` to `TODO`? I definitely agree we don't need it right now.

Reply:
I was thinking more in terms of making sure an initial field creation worked and error handling, but I think you're right and we can just leave as is.